[FEATURE] 회원탈퇴 기능 구현

JakePark929 opened the pull request:
## 📝 PR Summary

- 회원 탈퇴 로직이 추가되었습니다. ACTIVE 인 사용자 상태를 DELETE 로 변경합니다.
- `AuthBusiness` 회원 탈퇴 로직에 맞춰 사용자 로그인 로직이 수정되었습니다. 
> (기존) 로그인 시 ACTIVE 상태 사용자만 찾아서 반환 -> (변경) 모든 상태 사용자를 찾고 만약 탈퇴한 사용자이면 다시 활성화 상태로 변경
- `User` 도메인에도 코멘트 추가하였습니다. 불필요한 Setter 는 남발하지 않고 필요한 곳에만 어노테이션 하였습니다.

#### 🌲 Working Branch

feat/#9-user-delete

#### 🌲 TODOs

- [x] 회원 탈퇴 요청 시 회원 상태 변경
- [x] 탈퇴 후 재 가입 시 로직 구현 (`AuthBusiness` OAuth 로그인 로직 변경)

### Related Issues

resolve #9

### 📚 Remarks
이 외에도 회원 탈퇴 시 유저 정보를 언제까지 가지고 있을지, 
DELETE 에서 ACTIVE 로 변경한다면 따로 사용자에 알림이 필요한지,
정지 상태, 비활성 상태의 처리는 어떻게 할지 등 생각해볼 사항이 많이 있습니다만,
프로젝트 주제에서 많이 벗어나는 것 같아 우선 이 정도만 구현 하겠습니다.

다른 좋은 의견 있으시면 코멘트 주세요 :)

JakePark929 committed:
[#9] feat : 회원탈퇴 기능 구현

- 회원 탈퇴 로직이 추가되었습니다. ACTIVE 인 유저 상태를 DELETE 로 변경합니다.
- `AuthBusiness` 회원탈퇴 로직에 맞춰 유저 로그인 로직이 수정되었습니다. 기존에는 ACTIVE 상태 사용자만 찾아서 반환했으나, 전체 상태 회원을 찾고 만약 탈퇴한 회원이면 다시 활성화 상태가 됩니다.
- `User` 도메인에도 코멘트 추가하였습니다. 불필요한 Setter 는 남발하지 않고 필요한 곳에만 어노테이션 하였습니다.

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/auth/business/AuthBusiness.java:
```diff
@@ -59,7 +59,8 @@ public JwtResponse callbackGoogle(String authorizationCode) {
-            userDto = userService.getUserWithThrow(googleUserInfo.getSubId(), googleUserInfo.getEmail());
+            // 유저 조회하고 있으면 바로 반환, 만약 유저 상태가 ACTIVE 가 아니면 ACTIVE 로 바꾸고 사용자 반환
+            userDto = userService.getUserAndUpdateStatusWithThrow(googleUserInfo.getSubId(), googleUserInfo.getEmail());
         } catch (RuntimeException e) {
```
RuntimeException과 같이 큰 범위의 예외를 처리하게 된다면, 예측하지 못한 RuntimeException이 발생한 경우 어떻게 대처할 수 있을까요?

JakePark929 replied:
아, 이 부분은 저도 생각했는데 아직 Exception 에 대한 컨벤션이 명확하지 않아 
우선 CoreApiExcepion에서 상속받은 RuntimeException을 두었는데요.

CoreAPiException 의 ErrorType 또는 커스텀 Exception을 추가하여 해결할 수 있을 것 같습니다.
한 번 확인해보겠습니다~

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/user/controller/UserController.java:
```diff
@@ -14,15 +14,26 @@
     }
 
+    // 사용자 닉네임 변경
     @PutMapping("/set-nickname")
```
REST API 에서 동사 사용은 지양하고 있습니다! 
Put이라는 Http 메서드로 동작을 명시해주고 있으니, /nickname 으로 표현하는건 어떨까요?

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/user/service/UserService.java:
```diff
@@ -21,33 +21,65 @@ public class UserService {
+                userRepository.findFirstBySubIdAndEmailOrderByIdDesc(subId, email)
         ).orElseThrow(() -> new CoreApiException(ErrorType.USER_NOT_FOUND));
 
+        if (userEntity.getStatus() != UserStatus.ACTIVE) {
```
userEntity에서 status를 꺼내서 비교해야 할까요? 
UserEntity에게 역할을 위임할 수 있지 않을까요?

birdieHyun commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/user/service/UserService.java:
```diff
@@ -58,4 +90,20 @@ public void updateNickname(UserDto user, UserUpdateNicknameRequest request) {
         userRepository.save(userEntity);
     }
+
+    /**
```
이 부분은 팀별로 컨벤션인 부분이라 명확한 답은 없지만,

JPA의 dirty checking이 어떻게 동작하는지? update 할 때 save()를 호출하면 언제 save가 되고 언제 update가 되는지? save() 함수의 본래 목적 등과 같은걸 찾아보시면 좋을 것 같아요

JakePark929 replied:
JPA 의 dirty checking 개념은 저도 조금 있었는데 다른 팀원 분들을 위해 정의 부분 한번 다시 찾아 보았습니다!
JPA 의 dirty checking 은 엔티티의 필드가 변경될 시 JPA 가 이를 감지하여 트랜잭션이 커밋된 시점에 자동으로 반영 하는 것입니다. save() 함수는 엔티티의 Transient, Detached, Managed 3가지 상태에 따라 영속화와 DB 반영 여부를 결정하는 함수입니다~

donsonioc2010 replied:
저는 save(), saveAll() 등으로 명시를 해주는것도 좋더라구요, 직관적으로 저장을 한다는 알림의 의미로, 책을 읽듯 읽어지는게 좋아서요 🥰

JakePark929 committed:
[#9] refactor : User null 체크, 닉네임 변경 API 주소 변경 외 리뷰 내용 반영

- User 리포지터리에 Optional 객체를 이용해 null 체크를 진행했습니다.
- 사용자 활성화 여부를 외부에서 확인하지 않고 모델 내부에서 처리하도록 했습니다.
- 닉네임 변경 api 에 동사를 제외했습니다.

donsonioc2010 commented:
Conflict는 해결해놓으셔아할 것 같아요. Reminde드립니다. :)

JakePark929 committed:
Merge branch 'dev' into feat/#9-user-delete

JakePark929 commented:
로컬에서 리베이스 하다가 뭔가 꼬인 거 같아서요!
여기서 바로 처리했습니다!

확인 해주시면 감사하겠습니다~

donsonioc2010 commented on core/core-api/src/main/java/com/nerd/favorite18/core/api/user/service/UserService.java:
```diff
@@ -13,49 +13,91 @@
+     */
+    @Transactional
+    public void deleteUser(UserDto user) {
+        User userEntity = userRepository.findFirstByIdAndStatusOrderByIdDesc(user.getId(), UserStatus.ACTIVE)
+                .orElseThrow(() -> new CoreApiException(ErrorType.USER_NOT_FOUND));
+
+        userEntity.updateStatus(UserStatus.DELETE);
+
+        userRepository.save(userEntity);
+    }
```
이건 여담인데, 이부분은 자세히보니, 취향차이이긴한데, `userEntity`로 굳이 받을 필요 없이, `ifPresentOrElse()`로도 바로 끝낼 수 있는 코드같아보이네용 😃

JakePark929 replied:
좋은 의견 감사합니다! 👍 ifPresentOrElse() 메소드 추후 반영해 보겠습니다.

donsonioc2010 commented on storage/db-core/src/main/java/com/nerd/favorite18/storage/db/core/user/repository/UserRepository.java:
```diff
@@ -7,6 +7,6 @@
 
 public interface UserRepository  extends JpaRepository<User, Long> {
-    Optional<User> findBySubIdAndEmailAndStatusOrderByIdDesc(String subId, String email, UserStatus status);
+    Optional<User> findFirstBySubIdAndEmailOrderByIdDesc(String subId, String email);
     Optional<User> findFirstByIdAndStatusOrderByIdDesc(Long userId, UserStatus status);
```
이거는 물음적인 항목인데요, subId와 email로 묶는 항목이나, userId는 그야말로 단일로만 검색이 나올수밖에 없지 않을까 하는 쿼리 질의라고 생각하는데, OrderBy를 해야 하는 이유가 따로 있으신 걸까용?

JakePark929 replied:
네, 말씀하신 대로 사용자 쪽에서는 단일 검색이 예상되기 때문에 크게 의미는 없습니다.
다만, id 값이 키가 아니고, 단일 검색이 아닌 테이블의 경우 order by id 가 없다면 
데이터베이스는 조건에 맞는 행을 임의 순서로 반환하는 상황을 가정할 수 있기에
습관적으로 추가해 놓고는 합니다. :)

donsonioc2010 reviewed (approved):
궁금한거 코멘트만 남겨놨습니다.

따로 문제자체는 없는것 같아 Approve해놨습니다~

JakePark929 commented:
다들 리뷰 참여해주셔서 감사합니다. #20 PR 머지 하도록 하겠습니다.
